Add workgroupUniformLoad overload

beaufortfrancois opened the pull request:
Fixes https://github.com/gpuweb/gpuweb/issues/5124

alan-baker commented on wgsl/index.bs:
```diff
@@ -18767,7 +18767,27 @@ All synchronization functions [=shader-creation error|must=] only be invoked in
   <tr>
     <td>Description
-    <td>Returns the value pointed to by `p` to all invocations in the workgroup.
+    <td>Returns the non-atomic value pointed to by `p` to all invocations in the workgroup.
```
I'm not sure what non-atomic value is.
```suggestion
    <td>Returns the value pointed to by `p` to all invocations in the workgroup.
```

beaufortfrancois replied:
Done

alan-baker commented on wgsl/index.bs:
```diff
@@ -18767,7 +18767,27 @@ All synchronization functions [=shader-creation error|must=] only be invoked in
+    <td>`T` is a [=type/concrete=] [=integer scalar=] type.
+  <tr>
+    <td>Description
+    <td>Returns the atomic value pointed to by `p` to all invocations in the workgroup.
```
```suggestion
    <td>Atomically loads value pointed to by `p` and returns it to all invocations in the workgroup.
```

beaufortfrancois replied:
Done

alan-baker commented on wgsl/index.bs:
```diff
@@ -18767,7 +18767,27 @@ All synchronization functions [=shader-creation error|must=] only be invoked in
+      <xmp highlight=wgsl>
+        @must_use fn workgroupUniformLoad(p : ptr<workgroup, atomic<T>, read_write>) -> T
+      </xmp>
+  <tr>
+    <td>Parameterization
+    <td>`T` is a [=type/concrete=] [=integer scalar=] type.
```
I have a slight preference to dropping this since `atomic<T>` already implies this restriction and saves a potentially missed spec change if non-integral atomics ever get added.
```suggestion
```

beaufortfrancois replied:
Done

dneto0 reviewed (changes requested):
Please preserve the 'algorithm' attributes on the tr elements.

Otherwise looks fine.
(I had an earlier partial review that objected to the same things Alan pointed out. Thanks for fixing them.)

beaufortfrancois commented on wgsl/index.bs:
```diff
@@ -18756,7 +18756,7 @@ All synchronization functions [=shader-creation error|must=] only be invoked in
 
 <table class='data builtin'>
-  <tr algorithm="workgroupUniformLoad">
+  <tr>
```
Done

beaufortfrancois commented:
I've addressed your feedback. Thanks for the reviews folks!

> Please preserve the 'algorithm' attributes on the tr elements.
> 
> Otherwise looks fine. (I had an earlier partial review that objected to the same things Alan pointed out. Thanks for fixing them.)

Out of curiosity, how are `algorithm` attributes used in the spec?

dneto0 commented:
> 
> Out of curiosity, how are `algorithm` attributes used in the spec?

They provide context for `<var>` tags, which have a Bikeshed shorthand of vertical bars, like `|V|`.  Then in that context  the reader can click on a variable and all the other mentions of that variable will highlight with the same colour.

Here's the first row of the "vector access decomposition" rules where I've clicked on a bunch of the letters

<img width="822" alt="Screenshot 2025-04-03 at 11 05 23 AM" src="https://github.com/user-attachments/assets/94ddc258-1628-4adb-b5e1-ff93c9dc0bb9" />
It makes it much easier to follow what's going on.

In the particular case of workgroupUniformLoad there aren't variable definitions and uses, but I want to be consistent in the builtins tables and expressions tables.

alan-baker commented on wgsl/index.bs:
```diff
@@ -18775,6 +18775,23 @@ All synchronization functions [=shader-creation error|must=] only be invoked in
+      </xmp>
+  <tr>
+    <td>Description
+    <td>Atomically loads value pointed to by `p` and returns it to all invocations in the workgroup.
```
```suggestion
    <td>Atomically loads the value pointed to by `p` and returns it to all invocations in the workgroup.
```

alan-baker reviewed (approved):
Minor nit

beaufortfrancois commented:
Shall I merge or wait for other's approval as well?
